commons/zenoh-shm/src/posix_shm/array.rs

```diff
@@ -0,0 +1,119 @@
+//
+// Copyright (c) 2023 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+use std::{fmt::Display, marker::PhantomData, mem::size_of};
+
+use num_traits::{AsPrimitive, PrimInt, Unsigned};
+use stabby::IStable;
+use zenoh_result::{bail, ZResult};
+
+use super::segment::Segment;
+
+/// An SHM segment that is intended to be an array of elements of some certain type
+#[derive(Debug)]
+pub struct ArrayInSHM<ID, Elem, ElemIndex>
+where
+    rand::distributions::Standard: rand::distributions::Distribution<ID>,
+    ID: Clone + Display,
+{
+    inner: Segment<ID>,
+    _phantom: PhantomData<(Elem, ElemIndex)>,
+}
+
+unsafe impl<ID, Elem: Sync, ElemIndex> Sync for ArrayInSHM<ID, Elem, ElemIndex>
+where
+    rand::distributions::Standard: rand::distributions::Distribution<ID>,
+    ID: Clone + Display,
+{
+}
+unsafe impl<ID, Elem: Send, ElemIndex> Send for ArrayInSHM<ID, Elem, ElemIndex>
+where
+    rand::distributions::Standard: rand::distributions::Distribution<ID>,
+    ID: Clone + Display,
+{
+}
+
+impl<ID, Elem, ElemIndex> ArrayInSHM<ID, Elem, ElemIndex>
+where
+    rand::distributions::Standard: rand::distributions::Distribution<ID>,
+    ID: Clone + Display,
+    ElemIndex: Unsigned + PrimInt + 'static + AsPrimitive<usize>,
+    Elem: IStable<ContainsIndirections = stabby::abi::B0>,
+    isize: AsPrimitive<ElemIndex>,
+{
+    pub fn create(elem_count: usize, file_prefix: &str) -> ZResult<Self> {
+        if elem_count == 0 {
+            bail!("Unable to create SHM array segment of 0 elements")
+        }
+
+        let max: usize = ElemIndex::max_value().as_();
+        if elem_count - 1 > max {
+            bail!("Unable to create SHM array segment of {elem_count} elements: out of range for ElemIndex!")
+        }
+
+        let alloc_size = elem_count * size_of::<Elem>();
+        let inner = Segment::create(alloc_size, file_prefix)?;
+        Ok(Self {
+            inner,
+            _phantom: PhantomData,
+        })
+    }
+
+    pub fn open(id: ID, file_prefix: &str) -> ZResult<Self> {
+        let inner = Segment::open(id, file_prefix)?;
+        Ok(Self {
+            inner,
+            _phantom: PhantomData,
+        })
+    }
+
+    pub fn id(&self) -> ID {
+        self.inner.id()
+    }
+
+    pub fn elem_count(&self) -> usize {
+        self.inner.len() / size_of::<Elem>()
```

Comment:
Here `Elem` can not be a `ZST` otherwise this will be a division by `0`. See my suggestion above.